Passwordless-gdm for sssd-2-9

ikerexxe opened the pull request:
This is the implementation for the so called passwordless-gdm feature. The design page for this feature is available at https://github.com/SSSD/sssd.io/pull/79.

The original patch set was reviewed at https://github.com/SSSD/sssd/pull/8212 by Justin and Sumit. I've had done some minor modifications to those patches:
* Adapt the code to avoid merge/rebase conflicts.

As a reminder you can use https://copr.fedorainfracloud.org/coprs/ipedrosa/passwordles-gdm/ for testing and update authselect, mutter, gdm and gnome-shell packages. As for sssd, I'd use the build provided in this PR since it will contain the exact bits that will be shipped in RHEL.

authselect brings a new feature called `with-switchable-auth` that you should enable to use this feature. In addition, you should add the following configuration to sssd.conf:
```
[pam]
pam_json_services = gdm-switchable-auth
```

Known limitations:

* If EIdP and passkey are enabled at the same time for a given user, from the two of them only EIdP will be advertised (other methods aren't affected). Changing it will affect performance, so it's been decided to leave it as it is.
* Since https://github.com/SSSD/sssd/pull/7983 isn't merged, SSSD is unable to propagate the PIN request and PIN attempts left information to GDM. This is temporary and once this PR or the other one are merged, I'll update the PR to make these functionality available.

ikerexxe committed:
util: implement pam_get_response_data()

This API gets the selected response type data from the response_data
linked list. Includes unit tests.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>
Signed-off-by: Ray Strode <halfline@redhat.com>

ikerexxe committed:
Responder: generate JSON message for GUI

Implement a set of functions to check the available authentication
mechanisms and their associated data, and generate a JSON message with
it. This JSON formatted message will be consumed by apps that provide
GUI login (i.e. GDM). Currently, the implementation only takes into
account password and OAUTH2 mechanisms.

Include unit tests to check the implemented functions.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: call JSON message generation

Call JSON message generation function and fill the data structure
containing the response_data linked list.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
SSS_CLIENT: forward available auth JSON message

Forward the available authentication mechanisms and their associated
data message to the GUI login using a PAM conversation. Then, obtain the
reply and forward it to the responder, so that it can parse it.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>
Signed-off-by: Ray Strode <halfline@redhat.com>

ikerexxe committed:
Responder: unpack JSON reply from GUI

Implement a set of functions to unpack the JSON reply from the GUI.
Include unit tests to check the implemented functions.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: parse GUI reply

Parse GUI reply and set the appropriate data in `sss_auth_token`
structure.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: new option `pam_json_services`

This new option is used to enable the JSON protocol in the PAM responder
based on the PAM service file in use.

:config: A new option `pam_json_services` is now available to enable
         JSON protocol to communicate the available authentication
         mechanisms.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Test: adapt test_pam_srv to JSON message

Include JSON message where applies.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: check PAM service file for JSON protocol

Implement a function to check whether the PAM service file in use is
enabled for the JSON procotol. This helps us filter which applications
are compatible with this protocol.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: generate JSON message for smartcard

Implement a set of functions to retrieve the smartcard data and generate
the JSON message with it.

Implement new unit test and adapt the existing ones to take into account
the new data.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: parse reply for smartcard

Parse GUI reply for smartcard and set the appropriate data in
`sss_auth_token` structure.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: tune prompts in the GUI

Return `prompt_config` structure  in `pam_eval_prompting_config` to tune
the prompts from the SSSD config in the GUI.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
sss_client: add EIdP to prompt_config structure

Integration with GDM requests two prompts for EIdP so adding them to
prompt_config structure. In addition, implement all the functions needed
to manipulate the structure for these new prompts. Finally, add
unit-tests for the new functions.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
sss_client: modify smartcard in prompt_config structure

Integration with GDM requests two prompts for smartcard so modifying the
prompt_config structure. In addition, implement all the functions needed
to manipulate the structure for these new prompts. Finally, add
unit-tests for the new functions.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
util: implement pam_get_response_data_all_same_type()

This API gets all the elements with the selected response type data from
the response_data linked list. Includes unit tests.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: check return value for json_string()

It returns NULL on error, but this wasn't checked.

Fixes: ceeffa9e1 ("Responder: generate JSON message for GUI")

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: refactor JSON functions to reduce args

Several of the functions in `pamsrv_json` had lots of arguments and I'm
about to add more for the passkey authentication mechanism. Reduce these
arguments by creating a structure that will contain all these data.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: generate JSON message for passkey

Implement a set of functions to retrieve the passkey data and generate
the JSON message with it.

Implement new unit test and adapt the existing ones to take into account
the new data.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: parse reply for passkey

Parse GUI reply for passkey and set the appropriate data in
`sss_auth_token` structure.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
util: implement function to set passkey PIN

`sss_authtok_set_local_passkey_pin` provides a way to set the passkey
PIN in the authtok structure for local passkey authentication.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: extend smartcard JSON request message

Include the certificate data in the JSON messages to set it in the
authtok structure more easily.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: extend smartcard JSON reply message

Include the certificate data in the JSON message to set it in the
authtok structure more easily.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: make `decode_pam_passkey_msg()` public

This is needed by `pamsrv_json.c`, so let's make it public.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
krb5_child: advertise authentication methods

During the `preauthentication` phase krb5_child checks for the available
authentication methods for the given user, advertises them and the
process is kept alive. Once the state is change to `authentication` the
same krb5_child process processes the credentials and proceeds with the
authentication itself.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: update JSON message format

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: add `gdm-switchable-auth` to `pam_p11_allowed_services`
defaults

The `pam_p11_allowed_services` option now includes `gdm-switchable-auth`
as one of the default allowed PAM services for smartcard authentication.
The service was added alongside the other GDM-related services
(gdm-smartcard and gdm-password) for logical grouping.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
sss_client: prevent JSON auth during password change preauth

When a user's password expires after successful JSON authentication, the
fallback to traditional password change fails. Add
PAM_CLI_FLAGS_CHAUTHTOK_PREAUTH flag to distinguish password change
preauth from normal authentication preauth. When this flag is set, the
PAM responder skips JSON message generation and returns traditional
preauth data instead.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: change authentication mechanism detection

Use `pam_get_auth_types()` to detect the available mechanisms for a
user.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
man: clarify and fix `pam_json_services` compilation

Add a note to clarify that 2FA isn't supported in JSON protocol and fix
man page compilation for `pam_json_services` option.

:feature: Unified passwordless login in the GUI. SSSD now supports a
          rich authentication selection interface. Users can login with
          smartcards, passkey, External IdPs and passwords directly
          within the graphical user interface.
:packaging: SSSD now supports authentication mechanism selection through
            PAM using a JSON-based protocol. This feature enables
            passwordless authentication mechanisms in GUI login
            environments that support the protocol.
            Feature will be supported by GNOME Display Manager (GDM)
            starting with GNOME 50. While currently optimized for GNOME,
            the JSON protocol design allows for future support in other
            display managers.
            authselect is the recommended approach and will handle the
            necessary PAM stack modifications  automatically starting
            with version 1.7 through the new option `with-switch-auth`
            which provides a new PAM service called `switchable-auth`.
            Manual PAM configuration is also possible.
            For more technical details and implementation specifications,
            see the design documentation:
            https://github.com/SSSD/sssd.io/pull/79

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
krb5: port pre-authentication retry logic

Port the pre-authentication retry logic from the IPA provider to the
krb5 provider, making it available to all krb5-based authentication
flows.

Relates: 6c1272edf1 ("krb5: Add fallback password change support")
Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

sumit-bose committed:
krb5: fix OTP authentication

Resolves: https://github.com/SSSD/sssd/issues/8292
Reviewed-by: Justin Stephenson <jstephen@redhat.com>
(cherry picked from commit 60ba493e9664caecd9496901e8b9639c874b04ee)

ikerexxe committed:
krb5_child: fix OTP authentication for PAM stacked tokens

The `tokeninfo_matches()` function already handles PAM stacked tokens
correctly by processing them through the 2FA single path, so the
`answer_otp()` function should allow this token type to proceed.

Add SSS_AUTHTOK_TYPE_PAM_STACKED to the allowed authentication token
types in `answer_otp()` to restore previous functionality.

Fixes: 4cb99a248 ("krb5_child: advertise authentication methods").
Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>
Reviewed-by: Justin Stephenson <jstephen@redhat.com>
(cherry picked from commit df15165db43eeb380c2c2af0eb4492c647c266d5)

sumit-bose commented:
> Hi,
> 
> so far my tests went well, I opened [freeipa/freeipa#8075](https://github.com/freeipa/freeipa/pull/8075) to get some broader testing.
> 
> bye, Sumit

Ah, looks like I have to create a Fedora-43 build as well.

alexey-tikhonov committed:
KRB5_CHILD: allow `k5c_ccache_setup()` during SSS_PAM_PREAUTH

This should cover a case when a single execution of 'krb5_child'
handles both PREAUTH and AUTH

Resolves: https://github.com/SSSD/sssd/issues/8331
Reviewed-by: Iker Pedrosa <ipedrosa@redhat.com>
Reviewed-by: Sumit Bose <sbose@redhat.com>
(cherry picked from commit e2273e09a8e539fdb156bd1aba56c2d4210a7ad3)

ikerexxe committed:
krb5_child: fix enterprise principal parsing in keep-alive sessions

When keep-alive sessions transition between command types (e.g., from
SSS_PAM_PREAUTH to SSS_PAM_AUTHENTICATE), enterprise principal settings
were not being updated, causing parsing inconsistencies in complex AD
environments.

This change ensures that when the backend sends updated enterprise
principal settings for different command types, the principals are
correctly re-parsed with the appropriate flags, fixing UPN handling in
multi-domain AD environments.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>
Reviewed-by: Alejandro López <allopez@redhat.com>
Reviewed-by: Sumit Bose <sbose@redhat.com>
(cherry picked from commit dd3cd958d5b06e2429921ae46feeddfe137a1f7e)

ikerexxe commented on src/providers/krb5/krb5_ccache.c:
```diff
@@ -112,154 +112,6 @@ static errno_t find_ccdir_parent_data(TALLOC_CTX *mem_ctx,
-    return ret;
-}
-
-errno_t sss_krb5_precreate_ccache(const char *ccname, uid_t uid, gid_t gid)
```
I think I got this right now, but it'd be nice if you could take a look

ikerexxe commented on src/providers/krb5/krb5_child.c:
```diff
@@ -4204,14 +4207,13 @@ static krb5_error_code privileged_krb5_setup(struct krb5_req *kr,
         return ret;
     }
 
-    /* For ccache types FILE: and DIR: we might need to create some directory
```
I amended the patch

sumit-bose reviewed (approved):
Hi,

I agree with the latest changes. I run couple of  tests with the modified `sssctl user-checks` and they still worked as expected, ACK.

bye,
Sumit

spoore1 commented:
FYI, I've run tests with the updates in the copr repo here:
https://copr.fedorainfracloud.org/coprs/ipedrosa/passwordles-gdm/

All tests are passing on CentOS 9 Stream now:
```
======================================== short test summary info ========================================
PASSED tests/test_gdm.py::test_gdm__passkey_local_login_succeeds_with_pin (bare_ldap)
PASSED tests/test_gdm.py::test_gdm__passkey_login_succeeds_with_pin (bareipa)
PASSED tests/test_gdm.py::test_gdm__passkey_login_succeeds_when_pin_disabled (bareipa)
PASSED tests/test_gdm.py::test_gdm__passkey_login_succeeds_with_password_instead_of_pin (bareipa)
PASSED tests/test_gdm.py::test_gdm__passkey_login_succeeds_with_multiple_keys (bareipa)
PASSED tests/test_gdm.py::test_gdm__passkey_login_rejected_when_passkey_mapping_removed (bareipa)
PASSED tests/test_gdm.py::test_gdm__passkey_login_rejected_with_unregistered_mapping (bareipa)
PASSED tests/test_gdm.py::test_gdm__smartcard_login_succeeds_with_pin (bareipa)
PASSED tests/test_gdm.py::test_gdm__smartcard_login_fails_with_incorrect_pin (bareipa)
PASSED tests/test_gdm.py::test_gdm__smartcard_login_succeeds_with_certs_and_passkey (bareipa)
PASSED tests/test_gdm.py::test_gdm__xidp_login_rejected_for_invalid_password (gdm)
PASSED tests/test_gdm.py::test_gdm__xidp_login_rejected_when_user_disabled (gdm)
PASSED tests/test_gdm.py::test_gdm__xidp_user_is_forced_to_change_password_before_login (gdm)
PASSED tests/test_gdm.py::test_gdm__xidp_login_succeeds_and_gets_kerberos_ticket (gdm)
SKIPPED [1] tests/test_gdm.py:417: local no pin test not yet working
============================== 14 passed, 1 skipped in 2458.21s (0:40:58) ===============================
```

justin-stephenson reviewed (approved):
Ack from my side.

padelsbach committed:
pam: gate PAC indicator code on BUILD_SAMBA

Commit 1f680edad023c8c57343447b156f6b34696e8221 added ad_pac_common.c and
$(NDR_KRB5PAC_LIBS) to sssd_pam unconditionally. So when building --without-samba, sssd_pam fails to link with undefined references to ndr_pull_init_blob and ndr_pull_PAC_DATA.

This change qualifies those additions with `BUILD_SAMBA` so the PAC
indicator feature is compiled in only when samba support is enabled.

Reviewed-by: Sumit Bose <sbose@redhat.com>
Reviewed-by: Tomáš Halman <thalman@redhat.com>
(cherry picked from commit d0beceaa17b94fa07a2ee7af8165392b90e83d94)

sumit-bose committed:
krb5: restart krb5_child for Smartcard authentication

In contrast to other authentication methods for PKINIT some information
about the used Smartcard and certificate are already needed for the
pre-authentication step to trigger the MIT Kerberos PKINIT module to get
back the information if PKINIT is possible or not and if the Smartcard
can be used for authentication. If krb5_child is kept running between
the pre-authentication and the authentication step the information given
during pre-authentication is used if Smartcard authentication was
selected.

As long as only a single certificate is available there is no issue. But
if there are multiple certificates which all apply to the given mapping
and matching rules for the user trying to log in and the user can choose
a certificate for authentication the authentication might fail if the
certificate use during pre-authentication and the one selected by the
user differ. Before the change to keep krb5_child running for all
authentication methods this was not an issue since the fresh instance
started during the authentication step was using the certificate
selected by the user.

With this patch krb5_child is restart during the authentication step is
Smartcard authentication was selected.

Reviewed-by: Iker Pedrosa <ipedrosa@redhat.com>
Reviewed-by: Justin Stephenson <jstephen@redhat.com>
(cherry picked from commit f3a36bec2a6c9fe11076c8f4673775a0d4221ad1)

ikerexxe commented:
Latest changes contain a rebase to `sssd-2-9`, https://github.com/SSSD/sssd/commit/f3a36bec2a6c9fe11076c8f4673775a0d4221ad1 and partial https://github.com/SSSD/sssd/commit/d0beceaa17b94fa07a2ee7af8165392b90e83d94
